pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,723 @@
+				return cluster, err
+			}
+		}
+
+		if cluster.Spec.AKSConfig.Imported {
+			if cluster.Status.AKSStatus.UpstreamSpec == nil {
+				// non imported clusters will have already had upstream spec set
+				return e.setInitialUpstreamSpec(cluster)
+			}
+
+			if apimgmtv3.ClusterConditionPending.IsUnknown(cluster) {
+				cluster = cluster.DeepCopy()
+				apimgmtv3.ClusterConditionPending.True(cluster)
+				cluster, err = e.clusterClient.Update(cluster)
+				if err != nil {
+					return cluster, err
+				}
+			}
+		}
+
+		addNgMessage := "Cannot deploy agent without NodePools. Add a NodePool."
+		noNodePoolsOnSpec := len(cluster.Spec.AKSConfig.NodePools) == 0
+		noNodePoolsOnUpstreamSpec := len(cluster.Status.AKSStatus.UpstreamSpec.NodePools) == 0
+		if (cluster.Spec.AKSConfig.NodePools != nil && noNodePoolsOnSpec) || (cluster.Spec.AKSConfig.NodePools == nil && noNodePoolsOnUpstreamSpec) {
+			cluster, err = e.setFalse(cluster, apimgmtv3.ClusterConditionWaiting, addNgMessage)
+			if err != nil {
+				return cluster, err
+			}
+		} else {
+			if apimgmtv3.ClusterConditionWaiting.GetMessage(cluster) == addNgMessage {
+				cluster = cluster.DeepCopy()
+				apimgmtv3.ClusterConditionWaiting.Message(cluster, "Waiting for API to be available")
+				cluster, err = e.clusterClient.Update(cluster)
+				if err != nil {
+					return cluster, err
+				}
+			}
+		}
+
+		cluster, err = e.setTrue(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		if err != nil {
+			return cluster, err
+		}
+
+		// If there are no subnets it can be assumed that networking fields are not provided. In which case they
+		// should be created by the aks-operator, and needs to be copied to the cluster object.
+		if len(cluster.Status.AKSStatus.Subnets) == 0 {
+			subnets, _ := status["subnets"].([]interface{})
+			if len(subnets) != 0 {
+				// network field have been generated and are ready to be copied
+				virtualNetwork, _ := status["virtualNetwork"].(string)
+				subnets, _ := status["subnets"].([]interface{})
+				cluster = cluster.DeepCopy()
+
+				// change fields on status to not be generated
+				cluster.Status.AKSStatus.VirtualNetwork = virtualNetwork
+				for _, val := range subnets {
+					cluster.Status.AKSStatus.Subnets = append(cluster.Status.AKSStatus.Subnets, val.(string))
+				}
+				cluster, err = e.clusterClient.Update(cluster)
+				if err != nil {
+					return cluster, err
+				}
+			}
+		}
+		if cluster.Status.APIEndpoint == "" {
+			return e.recordCAAndAPIEndpoint(cluster)
+		}
+
+		if !cluster.Status.AKSStatus.UpstreamSpec.PublicAccess {
+			// Check to see if we can still use the public API endpoint even though
+			// the cluster has private-only access
+			serviceToken, mustTunnel, err := e.generateSATokenWithPublicAPI(cluster)
+			if mustTunnel != nil {
+				cluster = cluster.DeepCopy()
+				cluster.Status.ServiceAccountToken = serviceToken
+				return e.clusterClient.Update(cluster)
+			}
+			if err != nil {
+				return cluster, err
+			}
+		}
```

Comment:
Does AKS have a field that maps to public access? Is there a private setup? I saw that ip range restrictions can be set, is that what would be used to make a cluster private? If it is an actually option, have you tested a private setup?

Reply:
AKS has option for deploying prrivate cluster, I have to only re-think about setting up jump host or tunnel for connecting to it